### PR TITLE
Rework annotation ordering/optimisations

### PR DESCRIPTION
**Pull request by tecosaur**
There are some annoyances/discrepancies with annotation handling that I originally tried to fix in #53794, #53800 and #53801.

@LilithHafner was kind enough to have a look at these, and after a long chat we came to the conclusion that this isn't the right approach to take, and that the order with which annotations are applied should be given primacy over annotation ranges.

This PR supersedes the aforementioned PRs, and should make the way annotations are ordered more sensible.

See the commit messages for more details.

### CODE REVIEW

**Comment by tecosaur** [comment]
Ah, tests are failing because I missed a use of `Base.annotatedstring_optimize!` in StyledStrings' tests. I'll fix that and re-bump the stdlib shortly.

File: base/strings/annotated.jl
Comment (bvdmitri) [inline base/strings/annotated.jl:22]: These are not really typos but US vs UK way of writing.

Comment (tecosaur) [reply base/strings/annotated.jl:22]: Ah, not this again... Thanks for letting me know, I'll just drop this commit.

**Comment by tecosaur** [comment]
Ah yep, we did also discuss a need for more info in the docstrings. I'll see about adding some of the elaboration needed.

**Review by LilithHafner — COMMENTED** [review]
```julia
    for (region, annot) in str.annotations
        if region == fullregion
            push!(annotations, (firstindex(unannot):lastindex(unannot), annot))
        end
    end
    for offset in 0:len:(r-1)*len
        for (region, annot) in str.annotations
            if region != fullregion
                push!(annotations, (region .+ offset, annot))
            end
        end
    end
```
This is buggy in the new semantic because it re-orders annotations. Not necessarily blocking this PR, but an issue.

File: base/strings/annotated.jl
Comment (LilithHafner) [inline base/strings/annotated.jl:507]: This is O(n^2), but we can revise this function for perf later.

File: base/strings/annotated.jl
Comment (LilithHafner) [inline base/strings/annotated.jl:487]:
```suggestion
"""
    _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}, span::UnitRange{Int})

remove all annotations that fall within `span` by deleting ranges that are entirely contained, truncating
ranges that partially overlap, and splitting ranges that subsume span
"""
function _clear_annotations_in_region!(annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}, span::UnitRange{Int})
```
Documenting internals makes it easier for me to review. If this description is right, please accept the suggestion. Otherwise, I'll need to reevaluate my understanding of the function.

File: base/strings/annotated.jl
Comment (LilithHafner) [inline base/strings/annotated.jl:527]: This is also O(n^2) when it could be O(n), but perf can wait.

Comment (LilithHafner) [reply base/strings/annotated.jl:527]: And it will be O(n) in non-pathological cases.

File: base/strings/annotated.jl
Comment (LilithHafner) [inline base/strings/annotated.jl:512]: Does this function have unit tests?

**Review by LilithHafner — APPROVED** [review]
I think this should be merged and followup prs should address the performance of this PR's implementations and address the now-invalid-ness of the repeat function.

File: base/strings/annotated.jl
Comment (tecosaur) [reply base/strings/annotated.jl:527]: Yea, this is written with the non-pathological case in mind, and if you consider the case where few-annotation insertions are repeatedly made, individual insertions should be pretty much `O(1)` with a constant factor based on the average "checked annotation depth" / average number of annotations each insertion.

File: base/strings/annotated.jl
Comment (tecosaur) [reply base/strings/annotated.jl:512]: Yes, it's checked by the new block of tests at the end of `test/strings/annotated.jl`, under ``# The `_insert_annotations!` cautious-merging optimisation``.

File: base/strings/annotated.jl
Comment (tecosaur) [reply base/strings/annotated.jl:507]: Yep, and I don't think this should be in the "hot path" of any common use-cases :crossed_fingers:.

File: base/strings/annotated.jl
Comment (tecosaur) [reply base/strings/annotated.jl:487]: Mmmm, these functions _should_ all have docstrings. I'll see about adding a commit for this function and others.

**Comment by tecosaur** [comment]
I've just adjusted `repeat` to be consistent with the new semantic,
```julia
fullregion = firstindex(str):lastindex(str)
if allequal(first, str.annotations) && first(first(str.annotations)) == fullregion
    newfullregion = firstindex(unannot):lastindex(unannot)
    for (_, annot) in str.annotations
        push!(annotations, (newfullregion, annot))
    end
else
    for offset in 0:len:(r-1)*len
        for (region, annot) in str.annotations
            push!(annotations, (region .+ offset, annot))
        end
    end
end
```
this change has been folded into `2656884453 * Remove strong ordering of annotation ranges`

File: base/strings/annotated.jl
Comment (tecosaur) [reply base/strings/annotated.jl:496]: Docstrings added in `Document the semantics of annotation ordering`

**Comment by tecosaur** [comment]
Now that #54308 has been merged, we need to resolve the merge conflict and adjust the test introduced there.

**Comment by tecosaur** [comment]
After CI, I think this should be good to merge if that also sounds good by @LilithHafner :slightly_smiling_face:.

**Comment by LilithHafner** [comment]
1bed84f4ae6fe6fc9bb39612e274fd1630a51349 LGTM. I have not reviewed https://github.com/JuliaLang/julia/compare/1bed84f4ae6fe6fc9bb39612e274fd1630a51349..4c18472b3e56998f2784ecd5d9deea4cf5371850, but am open to suggestions to how I could do that without re-reviewing this PR from scratch.

**Comment by LilithHafner** [comment]
Okay, the diff in your comment also LGTM.

**Comment by IanButterworth** [comment]
@tecosaur I know you have a preference to not squash but for things that will get backported, it introduces unnecessary stumbling blocks in that automated process that would be better to avoid.

**Comment by tecosaur** [comment]
Thanks for mentioning that Ian, I do like to preserve more granular information when it seems sensible, but I wasn't aware that complicated the backporting process. I'll keep that in mind with the few remaining PRs I'm hoping to get backported :slightly_smiling_face:.

(As an aside, is the automated backporting doing something other than cherry picking?)

**Comment by IanButterworth** [comment]
Yeah the automatic script doesn't handle adding `-m`. If you wanted to help with adding that it's here https://github.com/KristofferC/Backporter/blob/master/backporter.jl

**Pull request merged**
